dftimewolf/lib/collectors/gsheets.py

```diff
@@ -0,0 +1,243 @@
+              secrets_path, self.SCOPES)
+          credentials = flow.run_console()
+
+        # Save the credentials for the next run
+        with open(credentials_path, 'w') as token_file:
+          token_file.write(credentials.to_json())
+
+    return credentials
+
+  def _BuildSheetsResource(self, credentials: Credentials) -> discovery.Resource:
+    """Builds a Google Sheets resource object to use to request logs.
+
+    Args:
+      credentials: Google API credentials
+
+    Returns:
+      A resouce object for interacting with the Google Sheets API.
+    """
+    resource = discovery.build('sheets', 'v4', credentials=credentials)
+
+    return resource
+
+  def _ValidateSpreadSheetId(self, spreadsheet: str) -> str:
+    """Extract the spreadsheet id if the input is a URL and validate that the ID
+    is in the correct format.
+
+    Args:
+      spreadsheet: ID or URL of the sheetspread,
+
+    Returns:
+      spreadsheet ID
+    """
+    spreadsheet_match = re.search(r'.*?([01][0-9A-Za-z_-]{20,}).*',
+                                     spreadsheet)
+    if not spreadsheet_match:
+      self.ModuleError(
+          'spreadsheet id is not in the correct format {0:s}.'.format(
+              spreadsheet),
+          critical=True)
+
+    return spreadsheet_match.group(1)
+
+  def _ParseSheet(self, spreadsheet_id: str, sheet_title: str) -> DataFrame:
+    """Parse sheet content
+
+      Args:
+      spreadsheet_id: ID of the spreadsheet to pull data from
+      sheets_title: Title of the sheet inside the spreadsheet to parse.
+
+      Returns:
+        Dataframe with entries from sheet inside the spreadsheet
+    """
+
+    df = pd.DataFrame()
+
+    resource = self._BuildSheetsResource(self._credentials)
+
+    sheet_content_result = resource.spreadsheets().values().get(
+        spreadsheetId=spreadsheet_id, range=sheet_title).execute()
+    values = sheet_content_result.get('values', [])
+
+    if not values:
+      self.logger.warning('No data found in sheet "{0:s}".'.format(sheet_title))
+      return None
+
+    df = pd.DataFrame(values[1:], columns=values[0])
+    df.replace('', 'NaN', inplace=True)
+    df.fillna('NaN', inplace=True)
+
+    # Removing white spaces from column names
+    df.rename(columns=lambda name: name.strip(), inplace=True)
+
+    for column in self._mandatory_columns:
+      if column not in df.columns:
+        self.logger.error(
+            'Mandatory column "{0:s}" was not found in sheet "{1:s}".'.format(
+                column, sheet_title))
+        self.logger.error('Please make sure all mandatory are present:')
+        self.logger.error(
+            '"message": String with an informative message of the event')
+        self.logger.error(
+            '"datetime": ISO8601 format for example: 2015-07-24T19:01:01+00:00')
+        self.logger.error(
+            '"timestamp_desc": String explaining what type of timestamp it is for example file created'
+        )
+        return None
```

Comment:
To keep modules decoupled from one another, wouldn't it be better to do this validation in the Timesketch section? I could imagine cases where we would like to tolerate a sheet 2 df collection where the intention is not to export it to Timesketch. 

Granted, it's better to to the validation as early as possible, so I think the best course of action here is to control this behavior with a flag specified in the recipe: if set, validate for timesketch, otherwise skip any header validation.

Reply:
Agree. Added an argument for that.